init typing

n-gao opened the pull request:
This PR adds typing support to seml. It also adds pyright to CI and pre-commit. For syncing the pre-commit dependencies we have a small script that will automatically pull the right dependencies from the pyproject.toml.

dfuchsgruber commented on pyproject.toml:
```diff
@@ -52,3 +53,7 @@ Changelog = "https://github.com/TUM-DAML/seml/releases/latest"
 seml = "seml.__main__:main"
+
+[tool.pyright]
+# TODO: Bump this to "strict" once we have type hints for everything.
```
Should this remain a TODO?

n-gao replied:
I don't know; do we ever want proper type checking? Tbh, it seems rather difficult.

dfuchsgruber commented on scripts/pre_commit_sync_pyright_deps.py:
I think this file may benefit from a header doctoring explaining that it copies the pyproject.tomls requirements to the pre-commit took

dfuchsgruber commented on src/seml/document.py:
```diff
@@ -0,0 +1,156 @@
+    description: str
+    executable: Required[str]
+    name: str
+    output_file: str
```
isn't that called output_dir?

dfuchsgruber commented on src/seml/document.py:
In general, I'd also greatly appreciate docstrings for all the TypedDicts. Specifically if they refer to the structure expected by an experiment file or the one stored in the MongoDB.

dfuchsgruber commented on src/seml/commands/start.py:
```diff
@@ -381,16 +388,16 @@ def start_local_job(
             )
 
-        logging.verbose(f'Running the following command:\n {cmd}')
+        logging.debug(f'Running the following command:\n {cmd}')
```
Ah, I guess that's why. But why go from verbose -> debug?

dfuchsgruber commented on src/seml/commands/start.py:
```diff
@@ -418,16 +425,16 @@ def start_local_job(
             shutil.rmtree(temp_dir)
-        if not use_stored_sources:
-            os.chdir(origin)
+        if wd is not None:
```
Obsolete with context manager.

dfuchsgruber commented on src/seml/commands/start.py:
```diff
@@ -746,7 +757,7 @@ def start_local_worker(
                 if (
                     debug_server
                     or output_to_console
-                    or logging.root.level <= logging.VERBOSE
```
Same thing, probably not relevant since you most likely have some rationale for switching levels...

dfuchsgruber commented on src/seml/commands/start.py:
```diff
@@ -802,7 +813,7 @@ def start_experiments(
         output_to_console = True
         srun = True
-        logging.root.setLevel(logging.VERBOSE)
+        logging.root.setLevel(logging.DEBUG)
```
verbose vs debug

dfuchsgruber commented on src/seml/commands/start.py:
```diff
@@ -1176,7 +1187,9 @@ def prepare_experiment(
     try:
-        job_info = get_slurm_jobs(get_current_slurm_job_id())[0]
+        slurm_id = get_current_slurm_job_id()
+        assert slurm_id is not None
```
maybe an error message alongside the assert?

dfuchsgruber commented on src/seml/database.py:
```diff
@@ -1,36 +1,48 @@
     if mongodb_config is None:
         mongodb_config = get_mongodb_config()
     db = get_database(**mongodb_config)
-    if suffix is not None and not collection_name.endswith(suffix):
```
why did you discard the suffix?

n-gao replied:
wasn't used anywhere in the code base

n-gao replied:
do you want to keep it?

dfuchsgruber commented on src/seml/database.py:
```diff
@@ -211,7 +231,9 @@ def build_filter_dict(filter_states, batch_id, filter_dict, sacred_id=None):
 
 
-def get_max_in_collection(collection, field: str):
+def get_max_in_collection(
```
Making return type `int` should solve an issue above?

n-gao replied:
I changed the function to accept an type

dfuchsgruber commented on src/seml/experiment/command.py:
```diff
@@ -117,7 +130,7 @@ def get_command_from_exp(
                 if not k.startswith(SETTINGS.NAMED_CONFIG.PREFIX)
             }
-            named_configs = interpolated[key_named_configs]
+            named_configs = cast(List[str], interpolated[key_named_configs])
```
I guess this cast can't be avoided, since we can not dynamically have a correct typing for the input dict for `resolve_config_interpolations` that has `key_named_configs` as a key of type `str`, or can we?

n-gao replied:
I am not sure what a good solution here is while not putting to many assumptions into the utility functions that technically are more general

dfuchsgruber commented on src/seml/experiment/config.py:
```diff
@@ -733,7 +747,7 @@ def _convert_value(value):
 
 
-def convert_values(val):
+def convert_values(val: Any):
```
If one wants to be super correct, one could provide overloading definitions to solve that if the input is of dict[K, V], the output will be of dict[K, Any]. (which avoids the cast in L.769?)

dfuchsgruber commented on src/seml/experiment/config.py:
```diff
@@ -800,10 +816,16 @@ def read_config(config_path: Union[str, Path]):
         slurm_list.append({})
 
-    return seml_dict, slurm_list, config_dict
+    return (
```
I really dislike casts, can we cast once into a type that resembles the config from yaml?

dfuchsgruber commented on src/seml/experiment/config.py:
```diff
@@ -853,7 +875,20 @@ def determine_executable_and_working_dir(config_path, seml_dict):
 
 
-def remove_prepended_dashes(param_dict):
+def remove_prepended_dashes(param_dict: dict[str, Any]) -> dict[str, Any]:
```
Probably should be generic of T and map from `dict[str, T]` to `dict[str, T]`

n-gao replied:
that's by definition a violation of the typing interface :D

dfuchsgruber commented on src/seml/experiment/mattermost_observer.py:
```diff
@@ -120,36 +124,43 @@ def __init__(
         )
 
-        self.run = None
+        self._run: dict[str, Any] | None = None
```
why `run` -> `_run`?

n-gao replied:
compatibility with the super class

dfuchsgruber commented on src/seml/experiment/mattermost_observer.py:
```diff
@@ -120,36 +124,43 @@ def __init__(
             )
             self.notify_on_heartbeat = True
 
+    @property
```
ok, what's the rationale behind making the run get only?

n-gao replied:
compatibility with the super class

dfuchsgruber commented on src/seml/utils/__init__.py:
```diff
@@ -247,10 +259,25 @@ def chunker(seq, size):
+# return the same type (for TypedDicts). Otherwise, we
+# want to return just a dict without any assumptions.
+@overload
+def merge_dicts(dict1: dict, dict2: dict) -> dict: ...  # type: ignore
```
Also no clue about that, sounds reasonable enough to me?

dfuchsgruber commented on src/seml/utils/__init__.py:
Maybe a side note: Its not really good practice to have so much functionality in an __init__.py, is it?

n-gao replied:
idk, we could split it up but I would call that out of scope for this PR

dfuchsgruber commented on src/seml/utils/__init__.py:
```diff
@@ -397,14 +423,14 @@ def format(self, record):
 
 class Hashabledict(dict):
-    def __hash__(self):
+    def __hash__(self):  # type: ignore - I don't think we can satisfy this. This is indeed a hack.
```
I'm pretty sure you can't, `Hashabledict` is, strictly speaking, not a proper subtype of `dict`, as it does not implement its interface for `hash`.

dfuchsgruber commented on src/seml/utils/__init__.py:
```diff
@@ -605,12 +633,12 @@ def wrapper(*args, **kwargs):
             return f(*args, **kwargs)
 
-        return wrapper
+        return wrapper  # type: ignore
```
This should be fixable?

n-gao replied:
technically no, but a cast is more explicit than a type ignore

dfuchsgruber commented on src/seml/utils/ssh_forward.py:
Did not review any functionality...

dfuchsgruber reviewed (changes requested):
Overall, I find this to be a strong pull request. While containing exceptionally many changes, it significantly increases the readability of the module while also fixing a so-far unnoticed bug.

If the authors can address some minor issues, I believe this to be a valuable contribution to `seml` and will approve of the pull request.

n-gao commented on src/seml/document.py:
```diff
@@ -0,0 +1,156 @@
+    status: Required[str]
+
+    # Set during runtime
+    artifacts: list
```
This would disturb the typing of pymongo and introduce more explicit type casts. I am in favor of having one type and just leave a comment for the runtime properties.

n-gao commented on src/seml/commands/add.py:
```diff
@@ -81,10 +84,75 @@ def filter_experiments(
+    # Now, check for duplicate configurations in the database.
+    documents = filter_experiments(collection, documents, use_hash=use_hash)
+    len_after = len(documents)
+    if len_after_deduplication != len_before:
```
I like this granularity since it allows one to figure out whether the duplicates are within the config itself or the DB.

dfuchsgruber replied:
So does changing the order though, right?

n-gao replied:
I dont get it :)

dfuchsgruber replied:
It's minor anyway: I just suggest code like:

```python
<deduping the added documents>
<printing the amounts of dups>
<deduping against the collection>
<printing the amount of dups>
```

That also helps having more expressive variable names. But if you don't want to change it, also fine.

n-gao commented on src/seml/commands/add.py:
```diff
@@ -271,14 +343,11 @@ def add_config_file(
 
-    batch_id = get_max_in_collection(collection, 'batch_id')
+    # Compute batch id
+    batch_id = cast(Optional[int], get_max_in_collection(collection, 'batch_id'))
```
I changed this such that get_max_in_collection takes a type as input which will call a runtime isinstance check.

dfuchsgruber replied:
Maybe an assert is cleaner then?

n-gao commented on src/seml/commands/add.py:
```diff
@@ -301,30 +370,35 @@ def add_config_file(
+        seml_config['source_files'] = upload_sources(seml_config, collection, batch_id)
+    del seml_config['use_uploaded_sources']  # type: ignore
+    # Now it is in the right format for the type checks
+    seml_config = cast(SemlDoc, seml_config)
```
I added new types in document.py, which represent the file content. Though these are very rudimentary, I recommend fleshing them out later instead of growing this PR.

dfuchsgruber replied:
Isn't this the kind of the point of the PR? Why do they need to be flashed out, or put differently, why does `read_config` not comply with the proposed `SemlDoc` already?

n-gao replied:
because the config in the yaml file differs from the config in the db, several keys only exist in either, I have added new types to capture this. But one PR doesn't have to finish the whole typing thing.

n-gao commented on src/seml/commands/add.py:
```diff
@@ -223,8 +294,9 @@ def assemble_slurm_config_dict(experiment_slurm_config: dict):
-        slurm_config['sbatch_options']
+    slurm_config['sbatch_options'] = cast(
+        SBatchOptions,
+        remove_prepended_dashes(cast(Dict[str, Any], slurm_config['sbatch_options'])),
```
well, obviously, the document before or after was a violation of the typeddict; otherwise, the function makes no sense. Does one over the other really make a difference?

n-gao commented on src/seml/commands/add.py:
```diff
@@ -301,30 +370,35 @@ def add_config_file(
-            ),
-            'config_unresolved': config_unresolved,
-        }
+        cast(
```
That would require another type cast since a proper experimentDoc requires many more keys than we have here.

n-gao commented on src/seml/commands/print.py:
```diff
@@ -628,7 +672,7 @@ def generate_queue_table(
     all_collections = set(db.list_collection_names())
 
-    collection_to_jobs = defaultdict(list)
+    collection_to_jobs: dict[tuple[str, str], list[dict[str, str]]] = defaultdict(list)
```
briefity, the other requires 2 more lines :D

n-gao commented on src/seml/commands/start.py:
```diff
@@ -183,7 +183,7 @@ def start_sbatch_job(
     prepare_args = ''
     if with_sources:
         prepare_args += ' --stored-sources-dir $tmpdir'
-    if logging.root.level <= logging.VERBOSE:
```
logging.verbose doesn't exists according to my type checkers, also can't find any reference about it

n-gao commented:
I changed a bunch of development and CI settings to cover Python and package versions better. I made the following changes:
* dropped support for Python 3.8, 3.9
* test in CI for the lowest and highest versions of each package
* I switched the CI completely to uv for faster CI times
* I also changed the instructions for seml development to use uv since that enables consistent, reproducible results with a lock file.

n-gao commented:
I also added very rudimentary support for non-conda environments by simply ignoring `site-packages` when uploading source files. Afaik, seml should just work fine without a conda environment since at the time of submitting a job, slurm will copy all environment variables anyway from the current shell. An issue arises if the virtual environment that submits the slurm job is not the same as the one is called `add`. I guess there are several options:
* Always copy all environment variables when adding an experiment and using these when submitting a job
* Supplying a path to the Python interpreter (this one could have unintended side effects as not all env vars might be set correctly, but I am not extremely deep into the pep about virtualenvs)
* Accepting conda environments or `activate` scripts from virtualenvs
